stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java

```diff
@@ -319,42 +375,94 @@ private void runInternal() throws Exception {
+          break;
+        }
+        case REMOTE_READ: // fall through
+        case CLUSTER_READ: {
+          remoteOrClusterRead();
+          break;
+        }
+        default:
+          throw new IllegalStateException("Unknown operation: " + mParameters.mOperation);
+      }
+    }
+
+    private void listFile() {
+      for (String nameJobWorkerDir : mJobWorkerDirNames) {
+        for (int testDirId = mThreadId; testDirId < mParameters.mNumDirs;
+             testDirId += mParameters.mThreads) {
+          String dirPath = String.format(TEST_DIR_STRING_FORMAT, mParameters.mLocalPath,
+                  nameJobWorkerDir, testDirId);
           File dir = new File(dirPath);
           dir.listFiles();
         }
-        return;
       }
+    }
 
-      for (int dirId = mThreadId; dirId < mParameters.mNumDirs; dirId += mParameters.mThreads) {
-        for (int fileId = 0; fileId < mParameters.mNumFilesPerDir; fileId++) {
-          mCurrentOffset = 0;
-          String filePath = String.format("%s/%d/%d", mParameters.mLocalPath, dirId, fileId);
-          while (!Thread.currentThread().isInterrupted()) {
-            if (isRead && CommonUtils.getCurrentMs() > mContext.getEndMs()) {
-              closeInStream();
-              return;
-            }
-            long ioBytes = applyOperation(filePath);
-
-            // Done reading/writing one file
-            if (ioBytes <= 0) {
-              break;
-            }
-            // Start recording after the warmup
-            if (CommonUtils.getCurrentMs() > recordMs) {
-              mThreadCountResult.incrementIOBytes(ioBytes);
-            }
+    private void writeOrLocalRead() throws Exception {
+      for (int testDirId = mThreadId; testDirId < mParameters.mNumDirs;
+          testDirId += mParameters.mThreads) {
+        for (int testFileId = 0; testFileId < mParameters.mNumFilesPerDir; testFileId++) {
+          String filePath = String.format(TEST_DIR_STRING_FORMAT + "/file-%d",
+              mParameters.mLocalPath, mBaseParameters.mId, testDirId, testFileId);
+          processFile(filePath, FuseIOOperation.isRead(mParameters.mOperation));
+        }
+      }
+      finishProcessingFiles();
+    }
+
+    private void remoteOrClusterRead() throws Exception {
+      for (int numJobWorkerDirProcessed = 0; numJobWorkerDirProcessed < mJobWorkerDirNames.size();
+          numJobWorkerDirProcessed++) {
+        // find which job worker directory to read
+        int indexCurrentJobWorkerDir = (numJobWorkerDirProcessed + mJobWorkerZeroBasedId)
+            % mJobWorkerDirNames.size();
+        // skip itself if the operation is remote read
+        if (indexCurrentJobWorkerDir == mJobWorkerZeroBasedId
+            && mParameters.mOperation == FuseIOOperation.REMOTE_READ) {
+          indexCurrentJobWorkerDir = (indexCurrentJobWorkerDir + 1) % mJobWorkerDirNames.size();
+        }
+        String nameCurrentJobWorkerDir = mJobWorkerDirNames.get(indexCurrentJobWorkerDir);
```

Comment:
Consider 4 job workers, mJobWorkerZeroBasedId = 1 Remote_Read
numJobWorkerDirProcessed = 0
indexCurrentJobWorkerDir = (0 + 1) / 4 = 1
indexCurrentJobWorkerDir == mJobWorkerZeroBasedId
            && mParameters.mOperation == FuseIOOperation.REMOTE_READ
indexCurrentJobWorkerDir = (1 + 1) % 4;  = 2
nameCurrentJobWorkerDir = ...dir2

numJobWorkerDirProcessed = 1
int indexCurrentJobWorkerDir = (1 + 1)
            % mJobWorkerDirNames.size(); = 2
String nameCurrentJobWorkerDir = dir2

did we read the same dir for twice?

Reply:
The directories at this level are created by job workers for distinguishing what files are written by whom. They are not `--num-dirs` given by users. And we agreed that the remote read is like cluster read, where each job worker will read from files created by every other job workers.

Reply:
Yeam but we read dirBelongingToJobWorker2 for twice, isn't it?

Reply:
Here https://github.com/Alluxio/alluxio/blob/5a14e865dda2248db30a692bc4056d4403db6045/stress/shell/src/main/java/alluxio/stress/cli/fuse/FuseIOBench.java#L436-L437 no job worker will go into the same dir.

Reply:
Ah I see...For remote read, I'm going to pretend it done reading the files it wrote without actually doing any reading.